src/browserlib/extract-references.mjs

```diff
@@ -68,14 +68,20 @@ function getExtractionRules(generator) {
  * @function
  * @private
  * @param {Node} node The DOM node to use as starting point
- * @param {String} name The sibling name to find
+ * @param {String} name The sibling name to find, "heading" to match any heading
+ * @param {Node} until The optional DOM sibling at which to stop no matter what
  * @return {Node} The next sibling with the given name, null if not found
  */
-function nextTag(node, name) {
+function nextTag(node, name, until) {
   let nextEl = node.nextElementSibling;
-  while(nextEl && nextEl.tagName !== name.toUpperCase()) {
+  while (nextEl && nextEl !== until &&
+      ((name === 'heading' && !nextEl.tagName.match(/^(H\d$|HGROUP)$/)) ||
```

Comment:
it may be more readable to having something like:
```js
const selector = name === "heading" ? "h1,h2,h3,h4,h5,h6,hgroup" : name;
``` 
and then in the loop condition 
```js
nextEl.matches(selector)
```